### PR TITLE
symbolic+imperative nn interface

### PR DESCRIPTION
**Pull request by piiswrong**
a high level NN interface that supports symbolic+imperative use. With a mixed Keras and pytorch flavor.

@jermainewang @pluskid @sxjscience @yajiedesign @tqchen @mli @madjam

### CODE REVIEW

**Comment by sxjscience** [comment]
Should we try to combine this with mx.mod.Module? Since we have written a lot of codes in Module.

File: example/autograd/data.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# pylint: skip-file
+""" data iterator for mnist """
+import sys
+import os
+# code to automatically download dataset
+curr_path = os.path.dirname(os.path.abspath(os.path.expanduser(__file__)))
+sys.path.append(os.path.join(curr_path, "../../tests/python/common"))
+import get_data
```
Comment (mli) [inline example/autograd/data.py]: suggest to add a function in test_utils.py, for example 

```python
def get_mnist():
    """Download and load the MNIST dataset

    Returns
    -------
    dict
        A dict containing
    """
    def read_data(label_url, image_url):
        with gzip.open(mx.test_utils.download(label_url)) as flbl:
            magic, num = struct.unpack(">II", flbl.read(8))
            label = np.fromstring(flbl.read(), dtype=np.int8)
        with gzip.open(mx.test_utils.download(image_url), 'rb') as fimg:
            magic, num, rows, cols = struct.unpack(">IIII", fimg.read(16))
            image = np.fromstring(fimg.read(), dtype=np.uint8).reshape(len(label), rows, cols)
            image = image.reshape(image.shape[0], 1, 28, 28).astype(np.float32)/255
        return (label, image)

    # changed to mxnet.io for more stable hosting
    # path='http://yann.lecun.com/exdb/mnist/'
    path='http://data.mxnet.io/data/mnist/'
    (train_lbl, train_img) = read_data(
        path+'train-labels-idx1-ubyte.gz', path+'train-images-idx3-ubyte.gz')
    (test_lbl, test_img) = read_data(
            path+'t10k-labels-idx1-ubyte.gz', path+'t10k-images-idx3-ubyte.gz')
    return {'train_data':train_img, 'train_label':train_lbl,
            'test_data':test_img, 'test_label':test_lbl}
```

then we can use NDArrayIter for them

File: python/mxnet/contrib/nn/layer.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+
+class Layer(object):
+    """Base class for all neural network layers and models.
+
+    Your models should subclass this class.
+
+    Layers can also contain other Layers, allowing you to nest them in a tree
+    structure. You can assign sublayers as regular attributes::
+        from mxnet import nn
```
Comment (mli) [inline python/mxnet/contrib/nn/layer.py]: there should be an empty line before from

Comment (mli) [inline python/mxnet/contrib/nn/layer.py]: also suggest to add >>> before each line, then we can have doctest to run all these commands

File: python/mxnet/contrib/nn/optim.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
+# coding: utf-8
+# pylint: disable=
```
Comment (mli) [inline python/mxnet/contrib/nn/optim.py]: i think both coding and pylint are not necessary

File: example/autograd/data.py (hunk 1 above)
Comment (piiswrong) [inline example/autograd/data.py]: This is temporary. I'll do something like pytorch.dataset later

File: python/mxnet/contrib/nn/layer.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+
+    def forward(self, x):
+        #pylint: disable=arguments-differ
+        for layer in self._children:
+            x = layer(x)
+        return x
+
+
+class SimpleLayer(Layer):
+    """SimpleLayer is a Layer that supports forwarding with both `Symbol` and `NDArray`.
```
Comment (ZihengJiang) [inline python/mxnet/contrib/nn/layer.py]: What's the difference between SimpleLayer and Layer? Layer cannot forward with NDArray?

**Comment by ZihengJiang** [comment]
Also, is `nn` a good name? User may feel confused about Layers in nn and operators in operator/nn folder? @piiswrong @mli

**Comment by sbodenstein** [comment]
@piiswrong: we are very interested in dynamic graph construction direction MXNet is going in. A question: how flexible (compared to pytorch) is this? For example, can complicated NLP examples be implemented (like https://devblogs.nvidia.com/parallelforall/recursive-neural-networks-pytorch/)?

**Comment by piiswrong** [comment]
@sbodenstein It should match pytorch feature by feature. If you see anything missing it should be easy to add it

**Comment by pluskid** [comment]
@sxjscience The existing module API can actually do low level forward backward easily, as a demo shown [here](https://github.com/dmlc/mxnet/blob/master/example/module/mnist_mlp.py#L45) for writing your own explicit training loop. But comparing to PyTorch's full interface, it is still a bit basic. I do think we have a lot to learn from PyTorch / Chainer API design by either extending module or have a new API.

File: example/autograd/mnist.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    metric = mx.metric.Accuracy()
+
+    for i in range(epoch):
+        train_data.reset()
+        for batch in train_data:
+            data = nn.utils.load_data(batch.data[0], ctx_list=ctx, batch_axis=0)
+            label = nn.utils.load_data(batch.label[0], ctx_list=ctx, batch_axis=0)
+            outputs = []
+            with ag.train_section():
+                for x, y in zip(data, label):
```
Comment (pluskid) [inline example/autograd/mnist.py]: Why are we having an inner loop? Is it iterating through the batch one example at a time? Is it used to demonstrate that we can do this or is it suggesting the user should do this?

File: python/mxnet/contrib/nn/layer.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    Your models should subclass this class.
+
+    Layers can also contain other Layers, allowing you to nest them in a tree
+    structure. You can assign sublayers as regular attributes::
+        from mxnet import nn
+        class Net(nn.Layer):
+            def __init__(self, prefix=None, params=None):
+                super(Net, self).__init__(prefix=prefix, params=params)
+                self.dense1 = nn.Dense(20, in_units=10, prefix=self.prefix+'dense1_')
+                self.dense2 = nn.Dense(20, in_units=20, prefix=self.prefix+'dense2_')
```
Comment (pluskid) [inline python/mxnet/contrib/nn/layer.py]: pytorch use metaclass magic to automatically name the sub-layer according to the variable name it get's assigned. So
```
self.dense1 = nn.Dense(...)
```
will automatically have a (nested) name `dense1`. However, I'm not sure this is actually a feature we want to have. It sometimes could confuse people. If you do something like
```
self.layers = [nn.Dense(...) for i in range(5)]
```
the meta-class will no longer be able to figure it out automatically, and not even be able to collect all the parameters in a layer.

File: python/mxnet/contrib/nn/layer.py [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+    prefix : str
+        Prefix acts like a name space. It will be prepended to the name of all Symbols and
+        Parameters created by this layer. Prefix should be unique within one network
+        to prevent name collisions.
+    params : ParameterDict or None
+        Manages Parameters of this Layer and sublayers. You can make two Layers share
+        parameter by passing the same dictionary to them. For example::
+            params = nn.ParameterDict(prefix='dense_')
+            dense1 = nn.Dense(20, in_units=10, prefix='dense1_', params=params)
+            dense2 = nn.Dense(20, in_units=10, prefix='dense2_', params=params)
```
Comment (pluskid) [inline python/mxnet/contrib/nn/layer.py]: What does this example mean? The two separate prefix `dense1_` and `dense2_` are used to name the layer, but their parameter will be named `dense_` XXX?

File: python/mxnet/contrib/nn/layer.py [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+    def __init__(self, prefix=None, params=None):
+        self._prefix = self._make_prefix(prefix)
+        self._name = self.prefix[:-1] if self.prefix.endswith('_') else self.prefix
+        if params is None:
+            params = ParameterDict(prefix=self.prefix)
+        self._params = params
+        self._children = []
+
+    def __setattr__(self, name, value):
+        """Automatically register sublayers."""
```
Comment (pluskid) [inline python/mxnet/contrib/nn/layer.py]: Maybe we are not going to make thing so complicated to support setting a `list` of layers or a `dict` of layers (sometimes people want to do that when there are a lot of inner layers or they are generated pragmatically). But maybe could be useful to detect this situation and print a warning that the layers might not be registered.

File: python/mxnet/contrib/nn/layer.py [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+        super(Layer, self).__setattr__(name, value)
```
Comment (pluskid) [inline python/mxnet/contrib/nn/layer.py]: Maybe we can use the `name` as a prefix if the layer being registered does not already explicitly have a prefix?

File: example/autograd/mnist.py (hunk 5 above)
Comment (piiswrong) [inline example/autograd/mnist.py]: This is doing data parallel on multiple gpus.
we can have a DataParallel util like pytorch but collecting the outputs to one gpu is not a good idea.

File: python/mxnet/contrib/nn/layer.py (hunk 7 above)
Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: yes

File: python/mxnet/contrib/nn/layer.py (hunk 9 above)
Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: I've considered this. But it gets complicated and confusing when you start to nest layers.

File: example/autograd/mnist.py (hunk 5 above)
Comment (pluskid) [inline example/autograd/mnist.py]: I see. Is it possible to hide the data-par mechanism like the original mxnet way? I guess in the imperative style API, the users always explicitly write for loops to handle time index or other dimensions. So it seems to be reasonable to always assume the 1st dimension is batch.

File: python/mxnet/contrib/nn/layer.py (hunk 9 above)
Comment (pluskid) [inline python/mxnet/contrib/nn/layer.py]: Hmm, I see. Yes, it should be recursive, and then becomes a bit complicated.

I checked the pytorch version. Their `Module` object do not have a name attached to it. The name is only used as a key when a module is nested in other modules and will be generated on the fly so that recursive key prefix is simple: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L440

For us, the module has its own name, and if we assign the same module as a submodule to two different places, it could be potentially conflicting names. So the situation is more complicated.

Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: Overall implicit registration on assign is pretty engineering. But it does make the interface easier to use.

**Comment by javelinjs** [comment]
Seems we are going to manually define a bunch of NN layers in python, e.g., Pooling, BatchNorm, etc. Can they be generated automatically from c lib, just like the functions for symbol? Other frontend languages can follow the same way to do code generation.

**Comment by tdomhan** [comment]
this is a super exciting development. I'm also wondering what the relationship to the existing APIs like the Module API will be in the long run?

File: python/mxnet/contrib/nn/layer.py (hunk 9 above)
Comment (jekbradbury) [inline python/mxnet/contrib/nn/layer.py]: Chainer used to use explicit registration at `__init__` time, and is moving to PyTorch's implicit registration on assign because it works much better with IDEs.

**Comment by jekbradbury** [comment]
This is awesome! My nvidia blogpost has one example of the kind of model that needs this sort of framework, but a more practically useful one that'll demonstrate the full power of this approach in MXNet is https://github.com/facebookresearch/clevr-iep. I think it would be surprisingly easy to port as a demonstration (one benefit of PyTorch not building in any training loop abstractions etc. is that there's less to port when you move it to another framework).

File: python/mxnet/contrib/nn/layer.py (hunk 9 above)
Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: @jekbradbury Could you elaborate?

Comment (jekbradbury) [inline python/mxnet/contrib/nn/layer.py]: Here's the issue/discussion https://github.com/pfnet/chainer/issues/1970 and PR https://github.com/pfnet/chainer/pull/2657
My comment on the issue was describing PyTorch's tentative API (it hadn't been released yet), and once it was released Chainer developers decided to adopt it for v2.

Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: is ide compatibility a huge issue? This kind of implicit behavior is great when it works but can cause lots of confusion when it doesn't.
I have two alternative solutions in mind:
`self.add('dense1', nn.Dense, 100, in_units=100)`
and
`self.dense1 = self.new('dense1', nn.Dense, 100, in_units=100)`
Any thoughts?

Comment (pluskid) [inline python/mxnet/contrib/nn/layer.py]: prefer `self.add('dense1', nn.Dense(100, in_units=100))`.

Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: The problem with this is you need to actually do
`nn.Dense(100, in_units=100, prefix=self.prefix+'dense1_', params=self.params.subdict('dense1_'))` which is a lot of repeated code.

Another possibility is to do something in __new__ so that layers created in __init__ are automatically passed `prefix=self.prefix+'dense1_', params=self.params.subdict('dense1_')`

Comment (jekbradbury) [inline python/mxnet/contrib/nn/layer.py]: The automatic registration approach has worked fairly well for PyTorch, except that we still get lots of beginning users confused about why they can't define modules inside 
lists: https://discuss.pytorch.org/t/generate-label-score-results-for-nllloss2d/2880/4

Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: @jekbradbury @pluskid 
How about this
```
        class Net(nn.Layer):
            def __init__(self, prefix=None, params=None):
                super(Net, self).__init__(prefix=prefix, params=params)
                with self:
                    self.dense1 = nn.Dense(...)
```
All layers created within the with section will be registered as self's child.

Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: or we can make users define`init` instead of `__init__` to hide the with:
```
class Layer(object):
  def __init__(self, *args, **kwargs):
    with self:
      self.__init__(*args, **kwargs)
  def init(self):
    pass
```

Comment (jekbradbury) [inline python/mxnet/contrib/nn/layer.py]: now it's magical again 😛 (but actually I prefer using `init` to `__init__` so that users don't have to call `super` and so that the magic is a little more obvious -- similar to how it's better for the forward function to be `forward` [PyTorch] rather than `__call__` [Chainer])

Comment (jekbradbury) [inline python/mxnet/contrib/nn/layer.py]: Actually the `with` approach is very clever; it solves the `list` vs `ModuleList` problem too!

Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: we can also call it `build` or `constructor` instead of `init` so that its even more obvious

Comment (piiswrong) [inline python/mxnet/contrib/nn/layer.py]: yeah that's the idea

Comment (pluskid) [inline python/mxnet/contrib/nn/layer.py]: I like this scope thing. Except that I would prefer not to hide it, but to require user to write it explicitly (so that they know what is happening and do not get confused when they accidentally overwrite `__init__` instead of `init`. So I'm suggesting maybe we could make it explicit by using a informative name for the scope such as 

```python
with self.collect_params():
    self.fcs = [nn.Dense(...) for i in range(3)]
    self.pred = nn.Dense(...)
```
`collect_params` could be replaced with a better name, but my point is to ask user to do it explicitly instead of hiding it. What do you guys think?

Comment (pluskid) [inline python/mxnet/contrib/nn/layer.py]: @jekbradbury as for `__call__` and `forward`. I think both are fine. What I do not like about pytorch is that it seems `__call__` does extra bookkeepings than `forward` and I do not what kind of consequence you will have by calling `forward` directly for user vs calling `__call__`. On the other hand, I think the Chainer way where `__call__` and `forward` are the same thing, sounds better, and less confusing.

**Comment by pluskid** [comment]
Several comments from previous conversations that I think might be good to post here for public discussions.

# Merging Metric and Loss
The current proposal serialize a list of metrics inside a loss object. An alternative proposal is to unify metric and loss. A `Metric` is an object that

- support `forward` that maps inputs to a scalar
- support `backward` that by default is NOP (i.e. zero gradient)

A `Loss` is a subclass of `Metric`, that has meaningful `backward` implementation.

When constructing module, a list of `Metric` can be given. During training and scoring, all the metrics will be `forward` evaluated. During training, all the `Loss` will be `backward` evaluated to provide initial gradient for back-prop.

So the effective loss is the sum of all `Loss` subclasses in the list metrics supplied.

# Layer and SimpleLayer
I think it might be good to rename Layer to BaseLayer, and SimpleLayer to Layer, as the user will mostly be interacting with `Layer`. 

The `simple_forward` function, which supports both NDArray forward and symbolic forward, might be renamed to `mixed_forward` or `generic_forward` (name coming from C++ generics). 

# Whether to provide a built-in `fit` function
The current `fit` implementation on Module is actually simple. However, it looks complicated because tooooooo many keyword arguments are added to that function. One way to potentially avoid this is to ask users to always write their own training loop, which should be fairly straightforward. 

To make common scenarios simple to use. We can provide a training loop code template, from which user can copy & paste and customize at will. Alternative, we could provide a simplified `fit` function, with only a few arguments, and state in the doc encouraging the user to write their own training loop if extension is needed.

File: example/autograd/mnist.py (hunk 5 above)
Comment (mli) [inline example/autograd/mnist.py]: it also confuses me. i thought you are iterating for example by example. probably rename `load_data` into something like `split_data_to_dev`.

File: example/autograd/resnet.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+            data = nn.utils.load_data(batch.data[0], ctx_list=ctx, batch_axis=0)
+            label = nn.utils.load_data(batch.label[0], ctx_list=ctx, batch_axis=0)
+            outputs = []
+            with ag.train_section():
+                for x, y in zip(data, label):
+                    z = net(x)
+                    loss = nn.loss.softmax_cross_entropy_loss(z, y)
+                    ag.compute_gradient([loss])
+                    outputs.append(z)
+            optim.step(batch.data[0].shape[0])
```
Comment (mli) [inline example/autograd/resnet.py]: change to `batch_size`

File: python/mxnet/contrib/nn/parameter.py [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+from collections import OrderedDict
+import numpy as np
+
+from ...base import mx_real_t
+from ... import symbol, ndarray, initializer, context
+from ...context import Context
+from ...contrib import autograd
+
+# pylint: disable= invalid-name
+tensor_types = (symbol.Symbol, ndarray.NDArray)
```
Comment (mli) [inline python/mxnet/contrib/nn/parameter.py]: you can change to 

```
tensor_types = (symbol.Symbol, ndarray.NDArray)  # pylint: disable=invalid-name
```
